File: nautical/io/cdata.py
from nautical.time import NauticalTime, TimeFormat
from nautical.log import get_logger
from nautical.location import Point
from nautical.noaa.buoy import BuoyData


log = get_logger()


def parse_winds(wind_data):
    '''Parse the wind information. The wind data includes
    a direction, speed in knots, as well as the gust speed
    
    :param wind_data: String containing the wind string
    :return: Dictionary containing the windspeed and gust information
    '''
    split_wind_data = wind_data.split()

    wind_speed = None
    gusts = None

    for data_point in split_wind_data:

        try:
            float_value = float(data_point)

            if wind_speed is None:
                wind_speed = float_value
            else:
                if gusts is None:
                    gusts = float_value
        except ValueError as error:
            log.debug(error)

    return {"wspd": wind_speed, "gst": gusts}


def parse_location(location_data):
    '''Parse the latitude and longitude values out of the
    the string that was passed in. The latitude and longitude
    should contain the NSEW strings describing their sign.

    :param location_data: String data contain latitude and longitude values
    :return: dictionary containing the location point (when valid)
    '''
    def _describe(data):
        '''Describe or determine lat vs lon, positivity, and
        the value that needs to be calculated. 
        '''
        _lookup = {
            "n": ("latitude", 1.0),
            "s": ("latitude", -1.0),
            "e": ("longitude", 1.0),
            "w": ("longitude", -1.0),
        }
        _data = data.lower()

        for key, value in _lookup.items():
            if key in _data:
                return value[0], value[1], _data.replace(key, '')
        return None, 0.0, None

    location_dict = {}  # base/default return
    split_location = [x for x in location_data.split() if x]

    latitude = None
    longitude = None

    for loc_data in split_location:
        ll_val, sign, val = _describe(loc_data)

        try:
            fval = float(val)

            if ll_val == "latitude":
                latitude = sign * fval
            elif ll_val == "longitude":
                longitude = sign * fval

        except (ValueError, TypeError) as error:
            log.debug(error)

    if None not in (latitude, longitude):
        location_dict["location"] = Point(latitude, longitude)

    return location_dict


def parse_time(time_data):
    '''Parse the month/day/year time out of the time
    data string from CDATA.

    :param time_data: String containing all time and date information
    :return: dictionary mm, dd, year, time where time is NauticalTime
    '''
    nautical_time = None
    month = None
    day = None
    year = None

    try:
        split_time_data = time_data.split()

        # date should be the first in the list
        date_data = split_time_data[0].split("/")
        month = int(date_data[0])
        day = int(date_data[1])
        year = int(date_data[2])

        # Get the minutes and hours out of the time in the list
        hour_min = split_time_data[1]
        hours = int(hour_min[:2])
        mins = int(hour_min[2:])
        nautical_time = NauticalTime(fmt=TimeFormat.HOUR_24)
        nautical_time.hours = hours
        nautical_time.minutes = mins
    except (IndexError, ValueError, TypeError) as error:
        log.debug(error)

    # output a dictionary containing the keys and values that are expected
    # during a normal web scrape of buoy information
    return {"time": nautical_time, "dd": day, "mm": month, "year": year}


# The alias is linked to the name contained in the BuoyData
aliases = {
    "gust": "gst",
    "wave height": "wvht",
    "significant wave height": "wvht",
    "dominant wave period": "dpd",
    "average wave period": "apd",
    "atmospheric pressure": "pres",
    # ptdy
    "air temperature": "atmp",
    "water temperature": "wtmp",
    "dew point": "dewp",        
    "salinity": "sal",
    "visibility": "vis",
    "tide": "tide",
    "swell wave heigh": "swh",
    "swell wave period": "swp",
    "wind wave height": "wwh",
    "wind wave period": "wwp",
    "wind speed": "wspd"
}

# The alias is actually multiple values, so let's use a function to parse
alias_func = {
    "winds": parse_winds,
    "location": parse_location
}


def parse_cdata(cdata):
    '''Parse the CDATA string that contains information about this 
    presumed buoy. The data here will be used as supplemental information
    as some buoys cannot be scraped online (no valid or useable data).

    :param cdata: String containing the CDATA or description element from kml
    :return: Dictionary containing all parsed fields.
    '''
    parsed_cdata_dict = {}

    # data is separated by line breaks, cut there
    xml_data = cdata.split("<br />")

    for xd_val in xml_data:

        # remove the tags
        elements = [x for x in xd_val.strip().replace("<b>", "").split("</b>") if x]
        # no elements found, don't bother searching
        if not elements:
            continue

        # format the string to remove colons and lowercase for searching
        key = elements[0].replace(":", "").lower()

        if key in alias_func:
            # all functions need to return a dictionary
            parsed_cdata_dict.update(alias_func[key](elements[1]))
        elif key in aliases:
            # The first element in the string will be the value
            # assumes default units
            try:
                parsed_cdata_dict[aliases[key]] = float(elements[1].split()[0])
            except ValueError as error:
                # skip the value if the value was not a float
                log.error(error)
        elif "utc" in key:
            parsed_cdata_dict.update(parse_time(key))

    # filter out the results that returned None
    return {k: v for k, v in parsed_cdata_dict.items() if v is not None}


def fill_buoy_with_cdata(buoy, cdata):
    '''Parse the CDATA string that contains information about this 
    presumed buoy. The data here will be used as supplemental information
    as some buoys cannot be scraped online (no valid or useable data).

    :param buoy: nautical.noaa.buoy.Buoy
    :param cdata: String containing the CDATA or description element from kml
    :return: Buoy object
    '''
    buoy_data = BuoyData()

    parsed_cdata = parse_cdata(cdata)

    # set the location for this station if found
    if "location" in parsed_cdata:
        loc = parsed_cdata.pop("location")
        buoy.location = loc

    # fill the buoy_data with the parsed data
    buoy_data.from_dict(parsed_cdata)
    buoy.present = buoy_data

File: pkg/io/cdata.go
package io

import (
	"encoding/json"
	"fmt"
	"strconv"
	"strings"

	loc "github.com/barbacbd/nautical/pkg/location"
	buoy "github.com/barbacbd/nautical/pkg/noaa/buoy"
	nt "github.com/barbacbd/nautical/pkg/time"
)

type LocationSign struct {
	LLType string
	Sign   float64
}

type NauticalDate struct {
	Month int
	Day   int
	Year  int
	Time  *nt.NauticalTime
}

var (
	aliasMap = map[string]string{
		"gust":                    "gst",
		"wave height":             "wvht",
		"significant wave height": "wvht",
		"dominant wave period":    "dpd",
		"average wave period":     "apd",
		"atmospheric pressure":    "pres",
		"air temperature":         "atmp",
		"water temperature":       "wtmp",
		"dew point":               "dewp",
		"salinity":                "sal",
		"visibility":              "vis",
		"tide":                    "tide",
		"swell wave height":       "swh",
		"swell wave period":       "swp",
		"wind wave height":        "wwh",
		"wind wave period":        "wwp",
		"wind speed":              "wspd",
	}

	locationMap = map[string]LocationSign{
		"n": {LLType: "latitude", Sign: 1.0},
		"s": {LLType: "latitude", Sign: -1.0},
		"e": {LLType: "longitude", Sign: 1.0},
		"w": {LLType: "longitude", Sign: -1.0},
	}
)

// removeEmpty is a helper function that removes empty string values from
// the original list
func removeEmpty(values []string) []string {
	var validSplitValues []string
	for _, str := range values {
		if str != "" {
			validSplitValues = append(validSplitValues, str)
		}
	}
	return validSplitValues
}

// ParseWinds converts the string wind information that can be retrieved from
// KML into the wind speed and gust information
func ParseWinds(windData string) (float64, float64, error) {
	windSpeed := 0.0
	windSet := false
	gust := 0.0
	gustSet := false

	splitWindData := strings.Split(windData, " ")

	for _, data := range splitWindData {
		value, err := strconv.ParseFloat(data, 64)
		if err == nil {
			if !windSet {
				windSpeed = value
				windSet = true
			} else if !gustSet {
				gust = value
				gustSet = true

				// break out of the loop as this is always last to be set
				break
			}
		}
	}

	if !windSet && !gustSet {
		return windSpeed, gust, fmt.Errorf("no wind speed or gust data parsed")
	}

	return windSpeed, gust, nil
}

// ParseLocation parses the latitude and longitude values out of string that can be retrieved
// from the KML
func ParseLocation(locationData string) (*loc.Point, error) {
	describe := func(data string) (string, float64, string, error) {
		lowerData := strings.ToLower(data)
		for key, value := range locationMap {
			if strings.Contains(lowerData, key) {
				return value.LLType, value.Sign, strings.Replace(lowerData, key, "", -1), nil
			}
		}
		return "", 0.0, "", fmt.Errorf("failed to find a matching lat/lon type")
	}

	var latitude float64
	latitudeSet := false
	var longitude float64
	longitudeSet := false
	splitData := strings.Split(locationData, " ")
	validSplitValues := removeEmpty(splitData)

	for _, value := range validSplitValues {
		llType, sign, strVal, err := describe(value)
		if err != nil {
			return nil, err
		}

		fval, err := strconv.ParseFloat(strVal, 64)
		if err != nil {
			return nil, err
		}

		switch {
		case llType == "latitude":
			latitude = fval * sign
			latitudeSet = true
		case llType == "longitude":
			longitude = fval * sign
			longitudeSet = true
		}
	}

	if !latitudeSet {
		return nil, fmt.Errorf("missing latitude value")
	} else if !longitudeSet {
		return nil, fmt.Errorf("missing longitude value")
	}

	return &loc.Point{Latitude: latitude, Longitude: longitude}, nil
}

// ParseTime parses the time data out of cdata retrieved from KML data.
func ParseTime(timeData string) (*NauticalDate, error) {
	splitTime := strings.Split(timeData, " ")

	dateData := strings.Split(splitTime[0], "/")
	month, err := strconv.Atoi(dateData[0])
	if err != nil {
		return nil, err
	}
	day, err := strconv.Atoi(dateData[1])
	if err != nil {
		return nil, err
	}
	year, err := strconv.Atoi(dateData[2])
	if err != nil {
		return nil, err
	}

	hourMin := splitTime[1]
	hours, err := strconv.Atoi(hourMin[:2])
	if err != nil {
		return nil, err
	}
	minutes, err := strconv.Atoi(hourMin[2:])
	if err != nil {
		return nil, err
	}

	nauticalTime := nt.NauticalTime{Minutes: minutes, Hours: hours, Format: nt.HOUR_24}
	return &NauticalDate{Month: month, Day: day, Year: year, Time: &nauticalTime}, nil
}

// ParseCData will parse the CDATA string that is pulled from the KML.
// On success, the returned Buoy structure should be altered as it will
// indicate that it was auto-filled. The Station and Description will need to be
// set to reflect valid Buoy information.
func ParseCData(cdata string) (*buoy.Buoy, error) {
	buoyData := buoy.BuoyData{}
	station := buoy.Buoy{
		Station:     "AutoFilled",
		Description: "Auto Filled Buoy From Parsed CData",
		Present:     &buoyData,
	}

	xmlData := strings.Split(cdata, "<br />")
	for _, xdVal := range xmlData {

		// Break up the data from the html line breaks
		trimmed := strings.TrimSpace(xdVal)
		trimmed = strings.Replace(trimmed, "<b>", "", -1)
		elements := strings.Split(trimmed, "</b>")

		if len(elements) == 0 {
			continue
		}
		key := strings.ToLower(strings.Replace(elements[0], ":", "", -1))

		switch {
		case key == "winds":
			windSpeed, gust, err := ParseWinds(elements[1])
			if err != nil {
				return nil, err
			}
			station.Present.WindSpeed = windSpeed
			station.Present.Gust = gust
		case key == "location":
			location, err := ParseLocation(elements[1])
			if err != nil {
				return nil, err
			}
			station.Location = loc.Point{Latitude: location.Latitude, Longitude: location.Longitude}
		case strings.Contains(key, "utc"):
			nauticalDate, err := ParseTime(key)
			if err != nil {
				return nil, err
			}
			station.Present.Time = nauticalDate.Time
			station.Present.Day = nauticalDate.Day
			station.Present.Month = nauticalDate.Month
			station.Present.Year = nauticalDate.Year
		default:
			if alias, ok := aliasMap[key]; ok {

				// Split all of the elements and grab the first one as it is the value we need
				validSplitValues := removeEmpty(strings.Split(elements[1], " "))
				val, err := strconv.ParseFloat(validSplitValues[0], 64)
				if err == nil {
					jsonStr := fmt.Sprintf("{\"%s\": %f}", alias, val)
					// nothing to do with the error for now
					json.Unmarshal([]byte(jsonStr), station.Present)
				}
			}
		}
	}

	return &station, nil
}
